### ISSUE TITLE
client is unstable When we improve concurrent

### ISSUE THREAD

**Issue by figoxu** [aerospike/aerospike-client-go#107]
I am using same code 、enviorment and the data which drive the test 。As the concurrent goes up ,the stability goes down .The API we test is 【 func (clnt *Client) Exists(policy *BasePolicy, key *Key) (bool, error)  】 The detail test result is as follow :
![qq 20151223095011](https://cloud.githubusercontent.com/assets/10055828/11969800/99181fee-a95a-11e5-8eab-b06ec8c86b25.png)
The program setup test data before we execute the test code .  The core test code is as follow : 
![qq 20151223095011](https://cloud.githubusercontent.com/assets/10055828/11969853/a5d57e92-a95b-11e5-82ae-2f80688e9b79.png)
@khaf  How can we improve the stability of client api ?

**Comment by GeertJohan** [comment]
@figoxu Aren't you just hitting the bottlenecks of your machine? If you have 500 goroutines and only 4, 8 or even 12 cores, the Go runtime will still schedule the goroutines across system threads. The runtime then switches between goroutines. Scheduling away from a goroutine can be done by the runtime when there is a network read. It'll continue on the goroutine later, but lets other goroutines run their code on the processor while the first one is waiting for data. The data from Aerospike could arrive in under a milisecond, but since there are so many goroutines in line and working on the processor, it takes some time before the scheduler can continue the first one.

So, long story short: I think your measurement is off because it doesn't only measure the time it takes to get answer from the server, but also time the goroutine spends in sleep.

**Comment by khaf** [comment]
As GeertJohan noted, too many goroutines will become counter productive, since scheduling so many goroutines will become a significant part of the time spent in your benchmark. 

In my own benchmarks, I came to the conclusion that 40-64 goroutines would deliver the maximum performance, and adding to that would degrade the performance.

This seems to be an inherent internal limitation of Go's runtime. To work around this, many people pool their goroutines and use a producer-worker idiom to control the number of goroutines and maximize performance.

**Comment by figoxu** [comment]
As the some concurrent ,the Go Aerospike Client Driver is a little bit fast than Java Aerospike Client Driver. It seems to be the bug of the Aerospike Server that "Exists" Method is unstable .After that , I using redis SortedSet , Storage same data , implements "Exists" Method with "ZScore" command of redis  On same code .Eventhough 500 goruntime , it works stability .Let's see the following picture:
![qq 20151223095011](https://cloud.githubusercontent.com/assets/10055828/12000236/6fa4dd78-ab0d-11e5-8b98-5d107fc863fc.png)

**Comment by khaf** [comment]
The maximum latency metric may include the time taken for establishing a connection + the operation, and it may happen in two situations:

1. When you are starting the benchmark without warm up and the connection pool is empty, in which the time for establishing new connections will be added to some operations and recorded.
2. When you run out of connections in the pool and need more. In these cases the time of establishing the connection will be recorded in the benchmark as well.

To make sure this is not contributing to the benchmark:

1. Make your connection pool big: `ClientPolicy.ConnectionQueueSize = 4096`
2. Prevent establishing new connections when the pool is full: `ClientPolicy.LimitConnectionsToQueueSize = true`
3. Warm up your benchmark. Before starting timing the benchmark, do a few thousand operations to make sure all needed connections are created.

If you have a code that you can share via a gist, please post the link here so that I can help you with finding the problem.

It would also be helpful if you could also add the number of each stat next to the percentage so that we can see the count as well.

Thanks.

**Comment by figoxu** [comment]
aerospike code


type Reader struct {
	ac *as.Client
	idChan chan string
	flag   chan bool
}
func (p *Reader) ReadLinesSlice(path string) {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("@err:", err)
		return
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		p.idChan <- scanner.Text()
	}
	p.flag <- true
}

func (p *Reader) do4ever(path,app string) {
	for {

		if err := p.do(path,app);err != nil {
			fmt.Println("[read error ] @err:", err)
		}
		fmt.Println("####finish####")
	}
}
func (p *Reader) do(path,app string) error {
	time.Sleep(time.Millisecond*time.Duration(sleepSeed*10))
	sleepSeed++
	go p.ReadLinesSlice(path)
	fmt.Println(".....start.....")
	for {
		select {
		case uuid := <-p.idChan:
			p.business(uuid, app)
		case <-p.flag:
			return nil
		default:
			time.Sleep(time.Millisecond * time.Duration(10))
		}
	}
	fmt.Println(".....finish.....")
	return nil
}

func (p *Reader) business(uuid,app string) error {
	key, err := dvKey(app, uuid)
	if err != nil {
		return err
	}
	begin := time.Now().UnixNano()
	_,e:= p.ac.Exists(AS_Read_POLICY, key)
	if e!=nil {
		fmt.Println("[read error ] @err:", e)
	}
	spend := time.Now().UnixNano() - begin

	atomic.AddInt64(&existCount,1)
	t :=spend/1e6
	if t<1{
		atomic.AddInt64(&spend_lowerThan_1ms,1)
	}else if t>1 && t<2 {
		atomic.AddInt64(&spend_greaterThan_1ms,1)
	}else if t>2 && t<4 {
		atomic.AddInt64(&spend_greaterThan_2ms,1)
	}else if t>4 && t<8 {
		atomic.AddInt64(&spend_greaterThan_4ms,1)
	}else if t>8 && t<16 {
		atomic.AddInt64(&spend_greaterThan_8ms,1)
	}else if t>16 && t<32 {
		atomic.AddInt64(&spend_greaterThan_16ms,1)
	}else if t>32 && t<64 {
		atomic.AddInt64(&spend_greaterThan_32ms,1)
	}else if t>64 && t<128 {
		atomic.AddInt64(&spend_greaterThan_64ms,1)
	}else if t>128 && t<256 {
		atomic.AddInt64(&spend_greaterThan_128ms,1)
	}else{
		atomic.AddInt64(&spend_greaterThan_256ms,1)
	}
	if existCount%10000 == 0 {
		log.Print("task execute:",existCount)
		log.Print("\t <1ms:",spend_lowerThan_1ms)
		log.Print("\t >1ms && <2ms :",spend_greaterThan_1ms)
		log.Print("\t >2ms && <4ms :",spend_greaterThan_2ms)
		log.Print("\t >4ms && <8ms :",spend_greaterThan_4ms)
		log.Print("\t >8ms && <16ms :",spend_greaterThan_8ms)
		log.Print("\t >16ms && <32ms :",spend_greaterThan_16ms)
		log.Print("\t >32ms && <64ms :",spend_greaterThan_32ms)
		log.Print("\t >64ms && <128ms :",spend_greaterThan_64ms)
		log.Print("\t >128ms && <256ms :",spend_greaterThan_128ms)
		log.Print("\t >256ms :",spend_greaterThan_256ms,"\n")
	}
	return e
}

------

type Reader struct {
	idChan chan string
	flag   chan bool
}

func (p *Reader) ReadLinesSlice(path string) {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("@err:", err)
		return
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		p.idChan <- scanner.Text()
	}
	p.flag <- true
}

func (p *Reader) do4ever(path, app string) {
	for {

		if err := p.do(path, app); err != nil {
			fmt.Println("[read error ] @err:", err)
		}
		fmt.Println("####finish####")
	}
}

func (p *Reader) do(path, app string) error {
	time.Sleep(time.Millisecond * time.Duration(sleepSeed*10))
	sleepSeed++
	go p.ReadLinesSlice(path)
	fmt.Println(".....start.....")
	for {
		select {
		case uuid := <-p.idChan:
			p.business(uuid, app)
		case <-p.flag:
			return nil
		default:
			time.Sleep(time.Millisecond * time.Duration(10))
		}
	}
	fmt.Println(".....finish.....")
	return nil
}

func (p *Reader) business(uuid, app string) error {
	appKey = fmt.Sprint("SYS_STATIC_DV_APP_a_", app)
	begin := time.Now().UnixNano()

	spend := int64(-1)
	c := REDIS_POOL.Get()
	defer c.Close()
	if _, err := redis.Int(c.Do("ZSCORE", appKey, uuid)); err != nil {
		if err == redis.ErrNil {
			fmt.Println("not found @app:", appKey, "@uuid:", uuid)
			spend = time.Now().UnixNano() - begin
		} else {
			log.Println("[read error] @error:", err)
			spend = -1
		}
	} else {
		spend = (time.Now().UnixNano() - begin)
	}
	if spend < 0 {
		return nil
	}



	atomic.AddInt64(&existCount, 1)
	t := spend / 1e6
	if t >= 0 && t < 1 {
		atomic.AddInt64(&spend_lowerThan_1ms, 1)
	} else if t > 1 && t < 2 {
		atomic.AddInt64(&spend_greaterThan_1ms, 1)
	} else if t > 2 && t < 4 {
		atomic.AddInt64(&spend_greaterThan_2ms, 1)
	} else if t > 4 && t < 8 {
		atomic.AddInt64(&spend_greaterThan_4ms, 1)
	} else if t > 8 && t < 16 {
		atomic.AddInt64(&spend_greaterThan_8ms, 1)
	} else if t > 16 && t < 32 {
		atomic.AddInt64(&spend_greaterThan_16ms, 1)
	} else if t > 32 && t < 64 {
		atomic.AddInt64(&spend_greaterThan_32ms, 1)
	} else if t > 64 && t < 128 {
		atomic.AddInt64(&spend_greaterThan_64ms, 1)
	} else if t > 128 && t < 256 {
		atomic.AddInt64(&spend_greaterThan_128ms, 1)
	} else {
		atomic.AddInt64(&spend_greaterThan_256ms, 1)
	}
	if existCount%10000 == 0 {
		log.Print("task execute:", existCount)
		log.Print("\t <1ms:", spend_lowerThan_1ms)
		log.Print("\t >1ms && <2ms :", spend_greaterThan_1ms)
		log.Print("\t >2ms && <4ms :", spend_greaterThan_2ms)
		log.Print("\t >4ms && <8ms :", spend_greaterThan_4ms)
		log.Print("\t >8ms && <16ms :", spend_greaterThan_8ms)
		log.Print("\t >16ms && <32ms :", spend_greaterThan_16ms)
		log.Print("\t >32ms && <64ms :", spend_greaterThan_32ms)
		log.Print("\t >64ms && <128ms :", spend_greaterThan_64ms)
		log.Print("\t >128ms && <256ms :", spend_greaterThan_128ms)
		log.Print("\t >256ms :", spend_greaterThan_256ms, "\n")
	}
	return nil
}

**Comment by figoxu** [comment]
redis code


Redis 

type Reader struct {
	idChan chan string
	flag   chan bool
}

func (p *Reader) ReadLinesSlice(path string) {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("@err:", err)
		return
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		p.idChan <- scanner.Text()
	}
	p.flag <- true
}

func (p *Reader) do4ever(path, app string) {
	for {

		if err := p.do(path, app); err != nil {
			fmt.Println("[read error ] @err:", err)
		}
		fmt.Println("####finish####")
	}
}

func (p *Reader) do(path, app string) error {
	time.Sleep(time.Millisecond * time.Duration(sleepSeed*10))
	sleepSeed++
	go p.ReadLinesSlice(path)
	fmt.Println(".....start.....")
	for {
		select {
		case uuid := <-p.idChan:
			p.business(uuid, app)
		case <-p.flag:
			return nil
		default:
			time.Sleep(time.Millisecond * time.Duration(10))
		}
	}
	fmt.Println(".....finish.....")
	return nil
}

func (p *Reader) business(uuid, app string) error {
	appKey = fmt.Sprint("SYS_XXXX_", app)
	begin := time.Now().UnixNano()

	spend := int64(-1)
	c := REDIS_POOL.Get()
	defer c.Close()
	if _, err := redis.Int(c.Do("ZSCORE", appKey, uuid)); err != nil {
		if err == redis.ErrNil {
			fmt.Println("not found @app:", appKey, "@uuid:", uuid)
			spend = time.Now().UnixNano() - begin
		} else {
			log.Println("[read error] @error:", err)
			spend = -1
		}
	} else {
		spend = (time.Now().UnixNano() - begin)
	}
	if spend < 0 {
		return nil
	}



	atomic.AddInt64(&existCount, 1)
	t := spend / 1e6
	if t >= 0 && t < 1 {
		atomic.AddInt64(&spend_lowerThan_1ms, 1)
	} else if t > 1 && t < 2 {
		atomic.AddInt64(&spend_greaterThan_1ms, 1)
	} else if t > 2 && t < 4 {
		atomic.AddInt64(&spend_greaterThan_2ms, 1)
	} else if t > 4 && t < 8 {
		atomic.AddInt64(&spend_greaterThan_4ms, 1)
	} else if t > 8 && t < 16 {
		atomic.AddInt64(&spend_greaterThan_8ms, 1)
	} else if t > 16 && t < 32 {
		atomic.AddInt64(&spend_greaterThan_16ms, 1)
	} else if t > 32 && t < 64 {
		atomic.AddInt64(&spend_greaterThan_32ms, 1)
	} else if t > 64 && t < 128 {
		atomic.AddInt64(&spend_greaterThan_64ms, 1)
	} else if t > 128 && t < 256 {
		atomic.AddInt64(&spend_greaterThan_128ms, 1)
	} else {
		atomic.AddInt64(&spend_greaterThan_256ms, 1)
	}
	if existCount%10000 == 0 {
		log.Print("task execute:", existCount)
		log.Print("\t <1ms:", spend_lowerThan_1ms)
		log.Print("\t >1ms && <2ms :", spend_greaterThan_1ms)
		log.Print("\t >2ms && <4ms :", spend_greaterThan_2ms)
		log.Print("\t >4ms && <8ms :", spend_greaterThan_4ms)
		log.Print("\t >8ms && <16ms :", spend_greaterThan_8ms)
		log.Print("\t >16ms && <32ms :", spend_greaterThan_16ms)
		log.Print("\t >32ms && <64ms :", spend_greaterThan_32ms)
		log.Print("\t >64ms && <128ms :", spend_greaterThan_64ms)
		log.Print("\t >128ms && <256ms :", spend_greaterThan_128ms)
		log.Print("\t >256ms :", spend_greaterThan_256ms, "\n")
	}
	return nil
}

**Comment by khaf** [comment]
After a week of trying to reproduce the problem, I still can't.

Just to make sure: Are the two databases installed on the same server? Are you sure you are not testing Redis on `localhost` and Aerospike on a different server?

**Comment by figoxu** [comment]
All of the program which include test client、aerospike、redis    are run in diferent host .
The redis which i use  is provide by ucloud . https://www.ucloud.cn/site/product/umem.html . 
we storage the aerospike server on the host which is provide by ucloud .https://www.ucloud.cn/site/product/uhost.html
